app/src/main/java/org/groundplatform/android/ui/datacollection/DataCollectionViewModel.kt

```diff
@@ -99,72 +115,94 @@ internal constructor(
-        savedStateHandle.getStateFlow(TASK_LOI_NAME_KEY, "")
-      } else
-      // LOI name pulled from LOI properties, if it exists.
-      flow {
-          locationOfInterestRepository.getOfflineLoi(surveyId, loiId)?.let {
-            val label = locationOfInterestHelper.getDisplayLoiName(it)
-            emit(label)
+    loadState
+      .map { (it as? LoadState.Ready)?.job?.name.orEmpty() }
+      .distinctUntilChanged()
+      .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5_000), "")
+
+  val loiName: StateFlow<String> =
+    if (loiId == null) {
+      savedStateHandle.getStateFlow(TASK_LOI_NAME_KEY, "")
+    } else {
+      surveyRepository.activeSurveyFlow
+        .filterNotNull()
+        .map { survey ->
+          withContext(ioDispatcher) {
+            locationOfInterestRepository
+              .getOfflineLoi(survey.id, loiId)
+              ?.let { locationOfInterestHelper.getDisplayLoiName(it) }
+              .orEmpty()
           }
-        })
-      .stateIn(viewModelScope, SharingStarted.Lazily, "")
+        }
+        .distinctUntilChanged()
+        .stateIn(viewModelScope, SharingStarted.Lazily, "")
+    }
 
   val loiNameDialogOpen: MutableState<Boolean> = mutableStateOf(false)
 
-  private val taskViewModels: MutableStateFlow<MutableMap<String, AbstractTaskViewModel>> =
-    MutableStateFlow(mutableMapOf())
+  init {
+    viewModelScope.launch {
+      _loadState.value = LoadState.Loading
+      val result = runCatching {
+        withTimeout(SURVEY_LOAD_TIMEOUT_MILLIS) {
+          val survey = surveyRepository.activeSurveyFlow.filterNotNull().first()
+          val job = survey.getJob(jobId) ?: error("couldn't retrieve job for $jobId")
+          val tasks =
+            if (isAddLoiFlow) job.tasksSorted else job.tasksSorted.filterNot { it.isAddLoiTask }
+          LoadState.Ready(survey = survey, job = job, tasks = tasks, surveyId = survey.id)
+        }
+      }
 
-  private val taskDataHandler = TaskDataHandler()
-  private val taskSequenceHandler = TaskSequenceHandler(tasks, taskDataHandler)
+      result
+        .onSuccess { ready ->
+          this@DataCollectionViewModel.job = ready.job
+          this@DataCollectionViewModel.tasks = ready.tasks
+          taskSequenceHandler = TaskSequenceHandler(tasks, taskDataHandler)
 
-  // Tracks the current task ID to compute the position in the list of tasks for the current job.
-  private val currentTaskId: StateFlow<String> = savedStateHandle.getStateFlow(TASK_POSITION_ID, "")
+          if (currentTaskId.value.isBlank()) {
+            savedStateHandle[TASK_POSITION_ID] = taskSequenceHandler.getValidTasks().first().id
+          }
 
-  init {
-    // Set current task's ID for new task submissions.
-    if (currentTaskId.value == "") {
-      savedStateHandle[TASK_POSITION_ID] = taskSequenceHandler.getValidTasks().first().id
+          _loadState.value = ready
+        }
+        .onFailure { e ->
+          Timber.e(e, "Failed to load survey/job")
```

Comment:
Maybe we should consider updating the error message to `Failed to initialize DataCollectionViewModel` to separate the error with the actual load errors.